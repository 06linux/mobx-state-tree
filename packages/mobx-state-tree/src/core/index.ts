export * from "./node"
export * from "./type"
export { createActionInvoker } from "./action"
export * from "./json-patch"
export * from "./mst-operations"
