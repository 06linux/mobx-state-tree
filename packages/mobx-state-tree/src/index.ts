// Fix some circular deps:
import "./core/type/type"
import "./core/node/object-node"
import "./core/node/scalar-node"

export { types, IModelType, IExtendedObservableMap } from "./types"

export {
    IType,
    ISimpleType,
    IComplexType,
    ISnapshottable,
    typecheckPublic as typecheck
} from "./core/type"

export * from "./core/mst-operations"
export { escapeJsonPath, unescapeJsonPath, IJsonPatch } from "./core/json-patch"
export {
    decorate,
    addMiddleware,
    IMiddlewareEvent,
    IMiddlewareHandler,
    IMiddlewareEventType
} from "./core/action"
export { process } from "./core/process"
export { isStateTreeNode, IStateTreeNode } from "./core/node"

export {
    applyAction,
    onAction,
    IActionRecorder,
    ISerializedActionCall,
    recordActions
} from "./middlewares/on-action"

export { createActionTrackingMiddleware } from "./middlewares/create-action-tracking-middleware"
