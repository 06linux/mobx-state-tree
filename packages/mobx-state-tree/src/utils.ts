import { isObservableArray, $mobx, getAtom } from "mobx"

/** @internal */
declare const global: any

/** @internal */
export const EMPTY_ARRAY: ReadonlyArray<any> = Object.freeze([])
/** @internal */
export const EMPTY_OBJECT: {} = Object.freeze({})

/** @internal */
export const mobxShallow =
    typeof $mobx === "string" ? { deep: false } : { deep: false, proxy: false }
Object.freeze(mobxShallow)

export type IDisposer = () => void

/** @internal */
export function fail(message = "Illegal state"): never {
    throw new Error("[mobx-state-tree] " + message)
}

/** @internal */
export function identity<T>(_: T): T {
    return _
}

/** @internal */
export function nothing(): null {
    return null
}

/** @internal */
export function noop() {}

// pollyfill (for IE) suggested in MDN:
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger
/** @internal */
export const isInteger =
    Number.isInteger ||
    function(value: any) {
        return typeof value === "number" && isFinite(value) && Math.floor(value) === value
    }

/** @internal */
export function isArray(val: any): boolean {
    return !!(Array.isArray(val) || isObservableArray(val)) as boolean
}

/** @internal */
export function asArray<T>(val: undefined | null | T | T[] | ReadonlyArray<T>): T[] {
    if (!val) return (EMPTY_ARRAY as any) as T[]
    if (isArray(val)) return val as T[]
    return [val] as T[]
}

/** @internal */
export function extend<A, B>(a: A, b: B): A & B
/** @internal */
export function extend<A, B, C>(a: A, b: B, c: C): A & B & C
/** @internal */
export function extend<A, B, C, D>(a: A, b: B, c: C, d: D): A & B & C & D
/** @internal */
export function extend(a: any, ...b: any[]): any
/** @internal */
export function extend(a: any, ...b: any[]) {
    for (let i = 0; i < b.length; i++) {
        const current = b[i]
        for (let key in current) a[key] = current[key]
    }
    return a
}

/** @internal */
export function extendKeepGetter<A, B>(a: A, b: B): A & B
/** @internal */
export function extendKeepGetter<A, B, C>(a: A, b: B, c: C): A & B & C
/** @internal */
export function extendKeepGetter<A, B, C, D>(a: A, b: B, c: C, d: D): A & B & C & D
/** @internal */
export function extendKeepGetter(a: any, ...b: any[]): any
/** @internal */
export function extendKeepGetter(a: any, ...b: any[]) {
    for (let i = 0; i < b.length; i++) {
        const current = b[i]
        for (let key in current) {
            const descriptor = Object.getOwnPropertyDescriptor(current, key)!
            if ("get" in descriptor) {
                Object.defineProperty(a, key, { ...descriptor, configurable: true })
                continue
            }
            a[key] = current[key]
        }
    }
    return a
}

/** @internal */
export function isPlainObject(value: any) {
    if (value === null || typeof value !== "object") return false
    const proto = Object.getPrototypeOf(value)
    return proto === Object.prototype || proto === null
}

/** @internal */
export function isMutable(value: any) {
    return (
        value !== null &&
        typeof value === "object" &&
        !(value instanceof Date) &&
        !(value instanceof RegExp)
    )
}

/** @internal */
export function isPrimitive(value: any): boolean {
    if (value === null || value === undefined) return true
    if (
        typeof value === "string" ||
        typeof value === "number" ||
        typeof value === "boolean" ||
        value instanceof Date
    )
        return true
    return false
}

/**
 * @internal
 * Freeze a value and return it (if not in production)
 */
export function freeze<T>(value: T): T {
    if (process.env.NODE_ENV === "production") return value
    return isPrimitive(value) || isObservableArray(value) ? value : Object.freeze(value)
}

/**
 * @internal
 * Recursively freeze a value (if not in production)
 */
export function deepFreeze<T>(value: T): T {
    if (process.env.NODE_ENV === "production") return value
    freeze(value)

    if (isPlainObject(value)) {
        Object.keys(value).forEach(propKey => {
            if (
                !isPrimitive((value as any)[propKey]) &&
                !Object.isFrozen((value as any)[propKey])
            ) {
                deepFreeze((value as any)[propKey])
            }
        })
    }

    return value
}

/** @internal */
export function isSerializable(value: any) {
    return typeof value !== "function"
}

/** @internal */
export function addHiddenFinalProp(object: any, propName: string, value: any) {
    Object.defineProperty(object, propName, {
        enumerable: false,
        writable: false,
        configurable: true,
        value
    })
}

/** @internal */
export function addHiddenWritableProp(object: any, propName: string, value: any) {
    Object.defineProperty(object, propName, {
        enumerable: false,
        writable: true,
        configurable: true,
        value
    })
}

/** @internal */
export function addReadOnlyProp(object: any, propName: string, value: any) {
    Object.defineProperty(object, propName, {
        enumerable: true,
        writable: false,
        configurable: true,
        value
    })
}

/** @internal */
export function remove<T>(collection: T[], item: T) {
    const idx = collection.indexOf(item)
    if (idx !== -1) collection.splice(idx, 1)
}

/** @internal */
export function registerEventHandler(handlers: Function[], handler: Function): IDisposer {
    handlers.push(handler)
    return () => {
        remove(handlers, handler)
    }
}

const prototypeHasOwnProperty = Object.prototype.hasOwnProperty
/** @internal */
export function hasOwnProperty(object: Object, propName: string) {
    return prototypeHasOwnProperty.call(object, propName)
}

/** @internal */
export function argsToArray(args: IArguments): any[] {
    const res = new Array(args.length)
    for (let i = 0; i < args.length; i++) res[i] = args[i]
    return res
}

/** @internal */
export function invalidateComputed(target: any, propName: string) {
    const atom = getAtom(target, propName) as any
    atom.trackAndCompute()
}

/** @internal */
export type DeprecatedFunction = Function & { ids?: { [id: string]: true } }
/** @internal */
export const deprecated: DeprecatedFunction = function(id: string, message: string): void {
    // skip if running production
    if (process.env.NODE_ENV === "production") return
    // warn if hasn't been warned before
    if (deprecated.ids && !deprecated.ids.hasOwnProperty(id)) {
        console.warn("[mobx-state-tree] Deprecation warning: " + message)
    }
    // mark as warned to avoid duplicate warn message
    if (deprecated.ids) deprecated.ids[id] = true
}
deprecated.ids = {}
