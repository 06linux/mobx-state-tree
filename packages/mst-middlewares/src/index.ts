export * from "./atomic"
export * from "./redux"
export * from "./simple-action-logger"
export * from "./time-traveller"
export * from "./undo-manager"
