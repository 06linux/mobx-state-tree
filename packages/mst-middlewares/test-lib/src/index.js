"use strict"
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p]
}
Object.defineProperty(exports, "__esModule", { value: true })
var atomic_1 = require("./atomic")
exports.atomic = atomic_1.default
__export(require("./redux"))
var simple_action_logger_1 = require("./simple-action-logger")
exports.simpleActionLogger = simple_action_logger_1.default
var time_traveller_1 = require("./time-traveller")
exports.TimeTraveller = time_traveller_1.default
var undo_manager_1 = require("./undo-manager")
exports.UndoManager = undo_manager_1.default
