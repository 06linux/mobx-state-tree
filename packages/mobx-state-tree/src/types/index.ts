// tslint:disable-next-line:no_unused-variable
import { IObservableArray, ObservableMap, IAction } from "mobx"
// tslint:disable-next-line:no_unused-variable
import { ISimpleType, IComplexType, IType } from "../core"
// tslint:disable-next-line:no_unused-variable
import { map, IExtendedObservableMap } from "./complex-types/map"
import { array } from "./complex-types/array"
import { identifier } from "./utility-types/identifier"
// tslint:disable-next-line:no_unused-variable
import { model, compose, IModelType } from "./complex-types/model"
import { reference } from "./utility-types/reference"
import { union } from "./utility-types/union"
import { optional } from "./utility-types/optional"
import { literal } from "./utility-types/literal"
import { maybe } from "./utility-types/maybe"
import { refinement } from "./utility-types/refinement"
import { frozen } from "./utility-types/frozen"
import { boolean, DatePrimitive, number, string, undefinedType, nullType } from "./primitives"
import { late } from "./utility-types/late"
import { enumeration } from "./utility-types/enumeration"

export { IModelType, IExtendedObservableMap }

export const types = {
    enumeration,
    model,
    compose,
    reference,
    union,
    optional,
    literal,
    maybe,
    refinement,
    string,
    boolean,
    number,
    Date: DatePrimitive,
    map,
    array,
    frozen,
    identifier,
    late,
    undefined: undefinedType,
    null: nullType
}
