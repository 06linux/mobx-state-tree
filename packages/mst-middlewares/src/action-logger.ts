import * as mst from "mobx-state-tree"

export function actionLogger(call: mst.IMiddlewareEvent, next: any) {
    const skip =
        (call.type === "action" && call.parentId !== 0) ||
        call.type === "flow_resume" ||
        call.type === "flow_resume_error"

    if (!skip)
        console.log(
            `[MST] #${call.rootId} ${call.type} - ${mst.getPath(call.context)}/${call.name}`
        )
    next(call)
}
