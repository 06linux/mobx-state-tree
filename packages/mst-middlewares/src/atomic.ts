import * as mst from "mobx-state-tree"

const atomic = mst.createActionTrackingMiddleware({
    filter: call => call.parentId === 0,
    onStart: call => mst.recordPatches(call.tree),
    onResume: (call, recorder) => recorder.resume(),
    onSuspend: (call, recorder) => recorder.stop(),
    onSuccess: (call, recorder) => {},
    onFail: (call, recorder) => recorder.undo()
})

export default atomic
