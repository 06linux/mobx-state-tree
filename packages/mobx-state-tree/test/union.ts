import { types, hasParent, tryResolve, getSnapshot } from "../src"
const createTestFactories = () => {
    const Box = types.model("Box", {
        width: types.number,
        height: types.number
    })
    const Square = types.model("Square", {
        width: types.number
    })
    const Cube = types.model("Cube", {
        width: types.number,
        height: types.number,
        depth: types.number
    })
    const Plane = types.union(Square, Box)
    const Heighed = types.union(Box, Cube)
    const DispatchPlane = types.union(
        snapshot => (snapshot && "height" in snapshot ? Box : Square),
        Box,
        Square
    )
    const Block = types.model("Block", {
        list: types.array(Heighed)
    })
    return { Box, Square, Cube, Plane, DispatchPlane, Heighed, Block }
}
if (process.env.NODE_ENV !== "production") {
    test("it should complain about multiple applicable types no dispatch method", () => {
        const { Box, Plane, Square } = createTestFactories()
        expect(() => {
            Plane.create({ width: 2, height: 2 })
        }).toThrowErrorMatchingSnapshot()
    })
}
test("it should have parent whenever creating or applying from a complex data structure to a model which has Union typed children", () => {
    const { Block, Heighed } = createTestFactories()
    const block = Block.create({
        list: [{ width: 2, height: 2 }]
    })
    const child = tryResolve(block, "./list/0")
    expect(hasParent(child)).toBe(true)
})
if (process.env.NODE_ENV !== "production") {
    test("it should complain about no applicable types", () => {
        const { Heighed } = createTestFactories()
        expect(() => {
            Heighed.create({ height: 2 })
        }).toThrowErrorMatchingSnapshot
    })
}
test("it should be smart enough to discriminate by keys", () => {
    const { Box, Plane, Square } = createTestFactories()
    const doc = types.union(Square, Box).create({ width: 2 })
    expect(Box.is(doc)).toEqual(false)
    expect(Square.is(doc)).toEqual(true)
})
test("it should discriminate by value type", () => {
    const Size = types.model("Size", {
        width: 0,
        height: 0
    })
    const Picture = types.model("Picture", {
        url: "",
        size: Size
    })
    const Square = types.model("Square", {
        size: 0
    })
    const PictureOrSquare = types.union(Picture, Square)
    const doc = PictureOrSquare.create({ size: { width: 0, height: 0 } })
    expect(Picture.is(doc)).toEqual(true)
    expect(Square.is(doc)).toEqual(false)
})
test("it should compute exact union types", () => {
    const { Box, Plane, Square } = createTestFactories()
    expect(Plane.is(Box.create({ width: 3, height: 2 }))).toEqual(true)
    expect(Plane.is(Square.create({ width: 3 }))).toEqual(true)
})
test("it should compute exact union types - 2", () => {
    const { Box, DispatchPlane, Square } = createTestFactories()
    expect(DispatchPlane.is(Box.create({ width: 3, height: 2 }))).toEqual(true)
    expect(
        DispatchPlane.is(
            Square.create({ width: 3, height: 2 } as any /* incorrect type, superfluous attr!*/)
        )
    ).toEqual(true)
})
test("it should use dispatch to discriminate", () => {
    const { Box, DispatchPlane, Square } = createTestFactories()
    const a = DispatchPlane.create({ width: 3 })
    expect(getSnapshot(a)).toEqual({ width: 3 })
})
