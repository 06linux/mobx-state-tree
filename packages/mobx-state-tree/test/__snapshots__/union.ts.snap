// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should complain about multiple applicable types no dispatch method 1`] = `
"[mobx-state-tree] Error while converting \`{\\"width\\":2,\\"height\\":2}\` to \`(Box | Square)\`:

    snapshot \`{\\"width\\":2,\\"height\\":2}\` is not assignable to type: \`(Box | Square)\` (Multiple types are applicable for the union (hint: provide a dispatch function)), expected an instance of \`(Box | Square)\` or a snapshot like \`({ width: number; height: number } | { width: number })\` instead."
`;
